File: apps/marketplace/apps/api/src/adapters/sms.ts
import { config } from '../lib/config.js';
import { logger } from '../lib/logger.js';

export interface SmsAdapter {
  sendOtp(phone: string, code: string): Promise<void>;
  verifyOtp(phone: string, code: string): Promise<boolean>;
}

class FakeSmsAdapter implements SmsAdapter {
  // In dev we don't actually send SMS — we log the code so devs can test the flow.
  // OTP storage/verification is delegated to OtpService (Redis or in-memory).
  async sendOtp(phone: string, code: string): Promise<void> {
    logger.info({ phone, code }, '[FakeSMS] OTP issued (dev only)');
  }
  async verifyOtp(): Promise<boolean> {
    // Verification handled by OtpService when using FakeSmsAdapter.
    return true;
  }
}

class TwilioVerifyAdapter implements SmsAdapter {
  async sendOtp(phone: string): Promise<void> {
    const url = `https://verify.twilio.com/v2/Services/${config.TWILIO_VERIFY_SID}/Verifications`;
    const auth = Buffer.from(`${config.TWILIO_ACCOUNT_SID}:${config.TWILIO_AUTH_TOKEN}`).toString('base64');
    const res = await fetch(url, {
      method: 'POST',
      headers: {
        Authorization: `Basic ${auth}`,
        'Content-Type': 'application/x-www-form-urlencoded',
      },
      body: new URLSearchParams({ To: phone, Channel: 'sms' }),
    });
    if (!res.ok) {
      logger.error({ status: res.status, body: await res.text() }, 'Twilio Verify send failed');
      throw new Error('Failed to send OTP');
    }
  }

  async verifyOtp(phone: string, code: string): Promise<boolean> {
    const url = `https://verify.twilio.com/v2/Services/${config.TWILIO_VERIFY_SID}/VerificationCheck`;
    const auth = Buffer.from(`${config.TWILIO_ACCOUNT_SID}:${config.TWILIO_AUTH_TOKEN}`).toString('base64');
    const res = await fetch(url, {
      method: 'POST',
      headers: {
        Authorization: `Basic ${auth}`,
        'Content-Type': 'application/x-www-form-urlencoded',
      },
      body: new URLSearchParams({ To: phone, Code: code }),
    });
    if (!res.ok) return false;
    const data = (await res.json()) as { status?: string };
    return data.status === 'approved';
  }
}

export const sms: SmsAdapter = config.TWILIO_FAKE
  ? new FakeSmsAdapter()
  : new TwilioVerifyAdapter();

File: apps/marketplace/apps/api-py/app/adapters/sms.py
"""SMS / OTP adapter — Twilio Verify in prod, in-memory fake in dev."""
from __future__ import annotations

import logging
import secrets
from abc import ABC, abstractmethod

import httpx

from app.config import settings

log = logging.getLogger(__name__)


class SmsAdapter(ABC):
    @abstractmethod
    async def send_otp(self, phone: str) -> str | None:
        """Returns the code only in dev so OtpService can store it."""

    @abstractmethod
    async def verify_otp(self, phone: str, code: str) -> bool:
        ...


class FakeSms(SmsAdapter):
    """Dev adapter — generates a 6-digit code and logs it."""

    async def send_otp(self, phone: str) -> str:
        code = f"{secrets.randbelow(1_000_000):06d}"
        log.info("[FakeSMS] OTP for %s -> %s (dev only)", phone, code)
        return code

    async def verify_otp(self, phone: str, code: str) -> bool:  # noqa: ARG002
        # Real verification handled by OtpService against its in-memory store.
        return True


class TwilioVerify(SmsAdapter):
    BASE = "https://verify.twilio.com/v2"

    def _auth(self) -> tuple[str, str]:
        assert settings.twilio_account_sid and settings.twilio_auth_token
        return settings.twilio_account_sid, settings.twilio_auth_token

    async def send_otp(self, phone: str) -> None:
        async with httpx.AsyncClient(timeout=10) as c:
            r = await c.post(
                f"{self.BASE}/Services/{settings.twilio_verify_sid}/Verifications",
                auth=self._auth(),
                data={"To": phone, "Channel": "sms"},
            )
            r.raise_for_status()
        return None

    async def verify_otp(self, phone: str, code: str) -> bool:
        async with httpx.AsyncClient(timeout=10) as c:
            r = await c.post(
                f"{self.BASE}/Services/{settings.twilio_verify_sid}/VerificationCheck",
                auth=self._auth(),
                data={"To": phone, "Code": code},
            )
            if r.status_code != 200:
                return False
            return r.json().get("status") == "approved"


sms: SmsAdapter = FakeSms() if settings.twilio_fake else TwilioVerify()
